src/__tests__/Validation.test.ts

```diff
@@ -94,3 +101,48 @@ describe('Rule', () => {
     );
   });
 });
+
+describe('validateObject', () => {
+  it('can validate over an object, returning the original with an attached error object containing messages', () => {
+    const validate = validateObject({
+      // required fields
+      firstName: [isNonEmptyString],
+      lastName: [isNonEmptyString],
+      email: [isNonEmptyString, isValidEmail],
+      // non-required fields
```

Comment:
It's a small thing, but not sure if I get the semantics of the "non-required fields" here. Doesn't this still set hasErrors to `true` in the case where it's `undefined` or `null`?

Reply:
that's a good question, it doesn't - but a form field would initialize to an empty string w or wo react

In a piece of arbitrary data you'd want an `isDefined` check